### ISSUE TITLE
Improve String diff for whitespace/newline character

### ISSUE THREAD

**Issue by joel-costigliola** [joel-costigliola/assertj-core#854]
#### Summary

When two strings difference is only about whitespaces or newlines it can be hard to see the difference. 

#### Example

```java
<add example here>
```

**Comment by xenoterracide** [comment]
this should track tabs, non breaking spaces, spaces, and newlines. For example if you have a tab at the beginning of a line, but are comparing it to 8 spaces, or if you have 3 spaces before the end of a line that aren't displayed with line wraps. Obviously, CRLF vs LF. Also knowing that someone has used a non breaking space unicode character in a copy and paste into a test can be a source of bane when refactoring.

**Comment by xenoterracide** [comment]
Also you may have to decide what to do if you have both whitespace differences and non whitespace mixed together. I would probably suggest always showing whitespace differences if they are present rather than showing if they are the only ones present. Probably harder to detect though.

**Comment by joel-costigliola** [comment]
Yeah, you are probably right, I still want to verify that the reported difference is still user friendly mixing whitespace diff and "normal" diff.

**Comment by xenoterracide** [comment]
hamcrest isEqual? does offer some support for this, so might look at its output for a queue, though I can't recall if it supports all of it off the top of my head, may play with it later.

**Comment by Fabinout** [comment]
Hi, I tried doing this.    
I'm not sure how I want the error message though, can I have a feedback ? 

```
"[Test] Displaying newline characters (\t, \n, ...)     
Expecting:     
  <"foo\n
">
to be equal to:
  <"foo\n
 \n
">
ignoring newline differences ('\r\n' == '\n')"
```

**Comment by joel-costigliola** [comment]
@Fabinout seems ok, can you add more examples: tab vs space,  non breaking spaces ...
@xenoterracide WDYT ? would you be able to provide test cases ?

**Comment by xenoterracide** [comment]
sorry it took me so long to respond... is there a class to put tests? with an existing test?, here's something I'm working on right now, but with hamcrest

		@Test
	public void preserveFirstParagraphIndentation()
	{
		String before
			= "<div>\n"
			+ "  &nbsp; \n\t" + CharacterUtils.NBSP
			+ " <p> Hello  </p> \n"
			+ " <p>"
			+ "<i>&nbsp;</i></p>"
			+ " &nbsp;</div>";

		String expected
			= "<div>\n    " + CharacterUtils.NBSP + "\n <p> Hello </p> \n <p><i> </i></p>  </div>";

		assertThat( HtmlSanitizer.cleanRichTextInput( before ), is( expected ) );
	}

the output is
```
Expected: is "<div>\n     \n <p> Hello </p> \n <p><i> </i></p>  </div>"
     but: was "<div>\n    \n\t  <p> Hello  </p> \n <p><i> </i></p>  </div>"
```

when Intellij Idea interprets that it creates a link for the diff tool, it  looks like

![screen shot 2017-06-16 at 2 54 25 pm](https://user-images.githubusercontent.com/5517/27242634-b8901cac-52a3-11e7-8e53-aebe1e752d01.png)

which is nice for everything but that unicode character. I think that wide characters in this case, maybe only whitespace ones should probably be rendered as 

```
Expected: is "<div>\n     \u00a0\n <p> Hello </p> \n <p><i> </i></p>  </div>"
     but: was "<div>\n    \n\t  <p> Hello  </p> \n <p><i> </i></p>  </div>"
```

I'm not sure if the Expected: is/ but: was is required for idea to give the "<Click to see difference>" link which I think should continue to work

**Comment by joel-costigliola** [comment]
@xenoterracide  Thanks for the feedback, AFAIK No PR has been submitted yet. 

Eclipse is able to show a visual comparison when the thrown `AssertionError` is actually a JUnit `ComparisonFailure` (I assume it's the same with Idea), we might get the result that we want by creating a `ComparisonFailure` with `actual` and `expected` showing clearly  tabs, non breaking spaces, spaces, and newlines.

@Fabinout have you got something that can be reviewed, even if not completed ?
